### PR TITLE
Polls frontend

### PR DESCRIPTION
**Pull request by Piwakk**
My great-grand-mother used to say: a good PR is like wine, if you don't wait for it long enough, it's tasteless.

### CODE REVIEW

File: frontend/src/index.tsx [hunk 1]
```diff
@@ -1,10 +1,10 @@
...
-import App from './App';
-import * as serviceWorker from './serviceWorker';
+import React from "react";
+import ReactDOM from "react-dom";
+import "./index.css";
+import App from "./App";
+import * as serviceWorker from "./serviceWorker";
 
-ReactDOM.render(<App />, document.getElementById('root'));
+ReactDOM.render(<App />, document.getElementById("root"));
```
Comment (abocquet) [inline frontend/src/index.tsx:7]: Out of the PR's scope, please roll back

File: frontend/src/models/associations/election.ts [hunk 2]
```diff
@@ -1,5 +1,5 @@
 import { User } from "../user";
-import {Association} from "./association";
+import { Association } from "./association";
```
Comment (abocquet) [inline frontend/src/models/associations/election.ts:3]: Out of the PR's scope, please roll back

File: frontend/src/serviceWorker.ts [hunk 3]
```diff
@@ -11,133 +11,134 @@
...
 
 const isLocalhost = Boolean(
-  window.location.hostname === 'localhost' ||
-    // [::1] is the IPv6 localhost address.
-    window.location.hostname === '[::1]' ||
-    // 127.0.0.1/8 is considered localhost for IPv4.
-    window.location.hostname.match(
-      /^127(?:\.(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}$/
-    )
+    window.location.hostname === "localhost" ||
```
Comment (abocquet) [inline frontend/src/serviceWorker.ts:14]: Out of the PR's scope, please roll back

File: frontend_legacy/package.json [hunk 4]
```diff
@@ -23,42 +23,42 @@
     "@ng-bootstrap/ng-bootstrap": "^4.2.2",
     "@ng-select/ng-select": "^2.20.5",
     "angular-markdown-editor": "^2.0.2",
-    "angular-notifier": "^4.1.1",
+    "angular-notifier": "^4.1.2",
```
Comment (abocquet) [inline frontend_legacy/package.json:26]: Out of the PR's scope, please roll back

File: frontend/src/components/polls/polls_table/PollEditModalUserForm.tsx [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                            });
+                            refetch({ force: true });
+                        }
+                    })
+                    .catch(error => {
+                        let message =
+                            "Erreur. Merci de réessayer ou de contacter les administrateurs si cela persiste.";
+                        let detail = error.response.data.detail;
+
+                        if (
```
[the remark this answers is not in the document]
Comment (Piwakk) [reply frontend/src/components/polls/polls_table/PollEditModalUserForm.tsx:57]: Problem is, several messages may correspond to the same error code, right?
For instance, here, there are three messages, and I guess they have either a 400 or a 403. So there should be at least two messages with the same error code

[the remark this answers is not in the document]
Comment (Piwakk) [reply frontend/src/components/polls/polls_table/PollEditModalUserForm.tsx:57]: And I think all the errors things are related to the [issue 143](https://github.com/Mines-Paristech-Students/Portail-des-eleves/issues/143). But this is a deeper work

[the remark this answers is not in the document]
Comment (Piwakk) [reply frontend/src/components/polls/polls_table/PollEditModalUserForm.tsx:57]: We need a real policy for handling errors like these.

File: frontend/src/components/polls/polls_table/PollEditModalAdminForm.tsx [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+            <SelectGroup
+                type="pills"
+                label="Statut"
+                name="state"
+                items={items}
+            />
+        );
+    }
+
+    function AdminCommentField() {
```
Comment (abocquet) [inline frontend/src/components/polls/polls_table/PollEditModalAdminForm.tsx:63]: Move as independent component outside of `PollEditModalAdminForm`, but not exported

File: frontend/src/components/polls/polls_table/PollEditModalUserForm.tsx [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+                    });
+                    refetch({ force: true });
+                }
+            })
+            .catch(error => {
+                let message =
+                    "Erreur. Merci de réessayer ou de contacter les administrateurs si cela persiste.";
+                let detail = error.response.data.detail;
+
+                if (detail === "You are not allowed to update this poll.") {
```
Comment (abocquet) [inline frontend/src/components/polls/polls_table/PollEditModalUserForm.tsx:45]: Check for 403 instead

File: frontend/src/components/polls/submit_polls/SubmitPoll.tsx [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+export const SubmitPoll = () => {
+    const newToast = useContext(ToastContext);
+
+    const [
+        questionPlaceholder,
+        choice0Placeholder,
+        choice1Placeholder
+    ] = getRandom([
+        ["Le portail…", "C’était mieux avant.", "C’est moins bien maintenant."],
```
Comment (abocquet) [inline frontend/src/components/polls/submit_polls/SubmitPoll.tsx:21]:
```suggestion
        ["Le portail…", "C’était pas mieux avant.", "C’est bien meilleur maintenant."],
```

**Comment by Piwakk** [comment]
Pagination is not included — for now.

File: frontend/src/components/polls/polls_table/PollEditModalAdminForm.tsx [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+                        level: ToastLevel.Success
+                    });
+                }
+            })
+            .catch(error => {
+                let message =
+                    "Erreur. Merci de réessayer ou de contacter les administrateurs si cela persiste.";
+                let detail = error.response.data.detail;
+
+                if (detail === "You are not allowed to update this poll.") {
```
Comment (abocquet) [inline frontend/src/components/polls/polls_table/PollEditModalAdminForm.tsx:113]:
```suggestion
                if (error.response.data.detail === "You are not allowed to update this poll.") {
```

File: frontend/src/components/polls/polls_table/PollEditModalAdminForm.tsx [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+                    newToast({
+                        message: "Sondage modifié.",
+                        level: ToastLevel.Success
+                    });
+                }
+            })
+            .catch(error => {
+                let message =
+                    "Erreur. Merci de réessayer ou de contacter les administrateurs si cela persiste.";
+                let detail = error.response.data.detail;
```
Comment (abocquet) [inline frontend/src/components/polls/polls_table/PollEditModalAdminForm.tsx:111]:
```suggestion
```

File: frontend/src/components/polls/polls_table/PollEditModalUserForm.tsx [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+                    });
+                    refetch({ force: true });
+                }
+            })
+            .catch(error => {
+                let message =
+                    "Erreur. Merci de réessayer ou de contacter les administrateurs si cela persiste.";
+                let detail = error.response.data.detail;
+
+                if (error.response.status === 403) {
```
Comment (abocquet) [inline frontend/src/components/polls/polls_table/PollEditModalUserForm.tsx:45]: Same

**Pull request merged**